### ISSUE TITLE
Linter for ex-info nil data

### ISSUE THREAD

**Issue by puredanger** [clj-kondo/clj-kondo#1314]
**Is your feature request related to a problem? Please describe.**
ex-info will error when passed a nil data map:

```clojure
user=> (ex-info "oops" nil)
Execution error (IllegalArgumentException) at user/eval1 (REPL:1).
Additional data must be non-nil.
``` 

It is quite easy however to pass a local into ex-info and forget to guard nil, particularly when the data map is the thing you're trying to error about:

```clojure
user=> 
(defn f [{:keys [a] :as m}] 
  (when (not a) (throw (ex-info "missing a" m)))
  a)
#'user/f
user=> (f {:a 1})
1
user=> (f nil)
Execution error (IllegalArgumentException) at user/f (REPL:1).
Additional data must be non-nil.
```

Workaround for that code:

```clojure
(defn f [{:keys [a] :as m}] 
  (when (not a) (throw (ex-info "missing a" (or m {}))))
  a)
```

**Describe the solution you'd like**
A linter that checks whether the ex-info data field is possibly nil.

**Describe alternatives you've considered**
Arguably, this is a Clojure problem where ex-info should be less annoying and I'm working on that side :) but in the meantime...

**Comment by puredanger** [comment]
Clojure ticket: https://clojure.atlassian.net/browse/CLJ-2640

**Comment by borkdude** [comment]
@puredanger I added some type information around `ex-info` which should catch _some_ cases:

![Screenshot 2021-07-06 at 20 09 21](https://user-images.githubusercontent.com/284934/124647315-155fe980-de96-11eb-9c88-e4c2f0a9df07.png)

![Screenshot 2021-07-06 at 20 05 13](https://user-images.githubusercontent.com/284934/124646901-894dc200-de95-11eb-8ae8-2cdfef2b7152.png)

but not yet this one. I guess that could be improved in the type inference.

![Screenshot 2021-07-06 at 20 08 17](https://user-images.githubusercontent.com/284934/124647193-ef3a4980-de95-11eb-9169-8718c6a9cdb3.png)

**Comment by borkdude** [comment]
I'll make another issue for that.

**Comment by puredanger** [comment]
Cool!
